Deserialization issue with Kryo 5 in parallel testing

Since updating to Kryo 5 we are experience a weird deserialization failure when running tests parallel on multiple Akka actor systems in the same JVM. We could not reproduce this while running tests sequential nor did this ever happen with Kryo 4.

```
Unable to find class: uth-store-syncer
Serialization trace:
value (scala.Some)
techInitiator (jaso.commons.security.SecurityCtx)
securityCtx (jaso.general.user.api.UserV0$DrctQryUserWithIdentity)
 com.esotericsoftware.kryo.KryoException: Unable to find class: uth-store-syncer
Serialization trace:
value (scala.Some)
techInitiator (jaso.commons.security.SecurityCtx)
securityCtx (jaso.general.user.api.UserV0$DrctQryUserWithIdentity)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:190)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:159)
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:691)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:118)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:124)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:729)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:125)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:124)
	at com.esotericsoftware.kryo.Kryo.readObjectOrNull(Kryo.java:780)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:134)
Caused by: java.lang.ClassNotFoundException: uth-store-syncer
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:604)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:416)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:184)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:159)
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:691)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:118)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:124)
```

The issue occurs deserializing a plain `String`, the field serializer tries to load a class by name with the content instead of just using that value. Any insight into what's going wrong would be very helpful.

> Ok that's really strange. Could you try to replicate the problem in a minimal test-case without Akka? It could create a couple of threads and serialize/deserialize the same object your original test uses. If that test passes, there has to be some issue in your current test setup.

I managed to write a test-case that reproduces this issue with plain Scala/Kryo:

```
import com.esotericsoftware.kryo.Kryo
import com.esotericsoftware.kryo.io.{Input, Output}
import com.esotericsoftware.kryo.util._
import org.objenesis.strategy.StdInstantiatorStrategy
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

import scala.concurrent._
import scala.concurrent.duration._


final case class Sample1(value: Option[String]) extends MySerializable {
  override def toString: String = "sample"
}

class ParallelSerializerTest extends AnyFlatSpec with Matchers {

  private val serializerPool = new Pool[Kryo](true, false, 8) {
    protected def create(): Kryo = {
      val kryo: Kryo = new Kryo(new DefaultClassResolver(), new MapReferenceResolver())
      kryo.setRegistrationRequired(false)
      kryo.setReferences(true)
      val instStrategy = kryo.getInstantiatorStrategy.asInstanceOf[DefaultInstantiatorStrategy]
      instStrategy.setFallbackInstantiatorStrategy(new StdInstantiatorStrategy())
      kryo.setInstantiatorStrategy(instStrategy)
      kryo
    }
  }

  private def serializeDeserialize(sample: Sample1): Unit = {
    val sKryo = serializerPool.obtain()
    val obuf = new Output(4096, -1)
    sKryo.writeClassAndObject(obuf, sample)
    val serialized = obuf.toBytes
    obuf.close()
    serializerPool.free(sKryo)

    val dKryo = serializerPool.obtain()
    val ibuf = new Input(serialized)
    val result = dKryo.readClassAndObject(ibuf)
    ibuf.close()
    serializerPool.free(dKryo)
    assert(result == sample)
  }

  it should "serialize and deserialize" in {
    implicit val dispatcher: ExecutionContextExecutor = ExecutionContext.global

    val results1 = for (i <- 0 to 10)
      yield Future {
        serializeDeserialize(Sample1(Some(s"test-$i")))
        serializeDeserialize(Sample1(Some(s"test-$i")))
        serializeDeserialize(Sample1(Some(s"test-$i")))
        serializeDeserialize(Sample1(Some(s"test-$i")))
      }

    Await.result(Future.sequence(results1), 10.seconds)
  }
}
```

This reproduces the same serialization issue we observed:
```
Unable to find class: est-2
Serialization trace:
value (scala.Some)
value (io.altoo.akka.serialization.kryo.Sample1)
com.esotericsoftware.kryo.KryoException: Unable to find class: est-2
Serialization trace:
value (scala.Some)
value (io.altoo.akka.serialization.kryo.Sample1)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:190)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:159)
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:691)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:118)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:124)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:729)
	at com.esotericsoftware.kryo.serializers.ReflectField.read(ReflectField.java:125)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:124)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:810)
	at io.altoo.akka.serialization.kryo.ParallelSerializerTest.serializeDeserialize(ParallelSerializerTest.scala:43)
	at io.altoo.akka.serialization.kryo.ParallelSerializerTest.$anonfun$new$3(ParallelSerializerTest.scala:54)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:671)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:430)
	at scala.concurrent.BatchingExecutor$AbstractBatch.runN(BatchingExecutor.scala:134)
	at scala.concurrent.BatchingExecutor$AsyncBatch.apply(BatchingExecutor.scala:163)
	at scala.concurrent.BatchingExecutor$AsyncBatch.apply(BatchingExecutor.scala:146)
	at scala.concurrent.BlockContext$.usingBlockContext(BlockContext.scala:107)
	at scala.concurrent.BatchingExecutor$AsyncBatch.run(BatchingExecutor.scala:154)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1425)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.ClassNotFoundException: est-2
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:416)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:184)
	... 24 more
```

--- a/src/com/esotericsoftware/kryo/util/DefaultGenerics.java
+++ b/src/com/esotericsoftware/kryo/util/DefaultGenerics.java
@@ -138,8 +138,10 @@ public final class DefaultGenerics implements Generics {
 
 	@Override
 	public Class resolveTypeVariable (TypeVariable typeVariable) {
-		for (int i = argumentsSize - 2; i >= 0; i -= 2)
-			if (arguments[i] == typeVariable) return (Class)arguments[i + 1];
+		for (int i = argumentsSize - 2; i >= 0; i -= 2) {
+			final Type arg = arguments[i];
+			if (arg == typeVariable || arg.equals(typeVariable)) return (Class)arguments[i + 1];
+		}
 		return null;
 	}
 

--- a/test/com/esotericsoftware/kryo/serializers/ParallelSerializationTest.java
+++ b/test/com/esotericsoftware/kryo/serializers/ParallelSerializationTest.java
@@ -1,0 +1,121 @@
+
+package com.esotericsoftware.kryo.serializers;
+
+import static org.junit.Assert.*;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.util.DefaultInstantiatorStrategy;
+import com.esotericsoftware.kryo.util.Pool;
+
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+import org.objenesis.strategy.StdInstantiatorStrategy;
+
+public class ParallelSerializationTest {
+
+	private final Pool<Kryo> pool = new Pool<Kryo>(true, false, 8) {
+		@Override
+		protected Kryo create () {
+			Kryo kryo = new Kryo();
+			kryo.setRegistrationRequired(false);
+			kryo.setReferences(true);
+			kryo.setInstantiatorStrategy(new DefaultInstantiatorStrategy(new StdInstantiatorStrategy()));
+			return kryo;
+		}
+	};
+
+	@Test
+	public void serializeAndDeserialize () {
+		IntStream.range(0, 100)
+			.parallel()
+			.forEach(it -> {
+				for (int i = 0; i < 10; i++) {
+					roundTrip(new TestClass(new GenericClass<>("test-" + it + "-" + i)));
+				}
+			});
+	}
+
+	private void roundTrip (Object target) {
+		byte[] serialized = serialize(target);
+		Object result = deserialize(serialized);
+		assertEquals(result, target);
+	}
+
+	private byte[] serialize (Object sample) {
+		Kryo kryo = pool.obtain();
+		Output out = new Output(4096, -1);
+		kryo.writeClassAndObject(out, sample);
+		byte[] serialized = out.toBytes();
+		out.close();
+		pool.free(kryo);
+		return serialized;
+	}
+
+	private Object deserialize (byte[] serialized) {
+		Kryo kryo = pool.obtain();
+		Input in = new Input(serialized);
+		Object result = kryo.readClassAndObject(in);
+		in.close();
+		pool.free(kryo);
+		return result;
+	}
+
+	static class TestClass {
+
+		private final GenericSuperClass<String> value;
+
+		public TestClass(GenericSuperClass<String> value) {
+			this.value = value;
+		}
+
+		public GenericSuperClass<String> value () {
+			return value;
+		}
+
+		@Override
+		public boolean equals (Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			TestClass that = (TestClass)o;
+			return Objects.equals(value, that.value);
+		}
+
+		@Override
+		public int hashCode () {
+			return Objects.hash(value);
+		}
+	}
+
+	static abstract class GenericSuperClass<T> {
+	}
+
+	static class GenericClass<T> extends GenericSuperClass<T> {
+		private final T value;
+
+		public GenericClass(T value) {
+			this.value = value;
+		}
+
+		public T value () {
+			return value;
+		}
+
+		@Override
+		public boolean equals (Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			GenericClass<?> genericClass = (GenericClass<?>)o;
+			return Objects.equals(value, genericClass.value);
+		}
+
+		@Override
+		public int hashCode () {
+			return value.hashCode();
+		}
+	}
+	
+}

